lld/ELF/Relocations.cpp

```diff
@@ -1374,8 +1379,8 @@ static unsigned handleTlsRelocation(RelType type, Symbol &sym,
             R_LOONGARCH_GOT_PAGE_PC, R_GOT_OFF, R_TLSIE_HINT>(expr)) {
     ctx.hasTlsIe.store(true, std::memory_order_relaxed);
     // Initial-Exec relocs can be optimized to Local-Exec if the symbol is
-    // locally defined.
-    if (execOptimize && isLocalInExecutable) {
+    // locally defined.  This is not supported on SystemZ.
+    if (execOptimize && isLocalInExecutable && config->emachine != EM_S390) {
```

Comment:
I investigated this in the weekend. Seems an unfortunate ABI issue when Local-Exec is unnecessarily inefficient
https://maskray.me/blog/2024-02-11-toolchain-notes-on-z-architecture#initial-exec-tls-model

Reply:
Yes, exactly.  Both Initial-Exec and Local-Exec sequences end up being LGRL+LGF, so any rewrite wouldn't actually be an optimization here.  Also, to rewrite into local-exec we'd have to allocate a constant pool entry slot somewhere, which is difficult for the linker.  Easier to just keep using the (already allocated) GOT slot.

Now, if we could use a more efficient Local-Exec sequence, this might be a different story.  Back when the ABI was defined, the architecture didn't yet have LGFI, so there wasn't really a way to load even a 32-bit constant without constant pool.  (B.t.w. are we certain that TLS offsets on a 64-bit platform can be restricted to 32 bits?)

Reply:
Thanks for the explanation. I've rewritten the paragraph to:

Optimizing the code sequence to local-exec is straightforward: changing the first instruction to `lgfi %r1, a@NTPOFF`.
However, LGFI (Load Immediate) is part of the extended-immediate facility (September 2005), introduced with System z9 109, unavailable when the ABI was defined.


> B.t.w. are we certain that TLS offsets on a 64-bit platform can be restricted to 32 bits?

We can. A 2GiB static TLS block almost assuredly won't work. It means that the dynamic loader needs to allocate 2GiB thread stack upfront for each new thread. The memory use is just not affordable.
https://github.com/llvm/llvm-project/issues/77128